File: backend/app/schemas/workspace.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class DatasetStats(BaseModel):
    images: int
    videos: int
    annotations: int
    classes: int
    labeled: int
    verified: int
    labeled_percent: float
    verified_percent: float


class DatasetSplits(BaseModel):
    train: int
    val: int
    test: int


class DatasetItem(BaseModel):
    id: int
    project_id: int | None
    name: str
    version: str
    topic: str
    task: str
    status: str
    format: str
    source_path: str
    storage_path: str
    created_at: datetime
    updated_at: datetime
    stats: DatasetStats
    splits: DatasetSplits


class DatasetMediaItem(BaseModel):
    id: str
    name: str
    path: str
    kind: str
    split: str
    annotated: bool
    verified: bool
    tags: list[str] = Field(default_factory=list)
    preview_url: str | None = None
    width: int | None = None
    height: int | None = None


class DatasetAnnotationShape(BaseModel):
    class_id: int
    x_center: float = Field(ge=0.0, le=1.0)
    y_center: float = Field(ge=0.0, le=1.0)
    width: float = Field(ge=0.0, le=1.0)
    height: float = Field(ge=0.0, le=1.0)


class DatasetAnnotationPayload(BaseModel):
    path: str = Field(min_length=1)
    annotations: list[DatasetAnnotationShape] = Field(default_factory=list)
    verified: bool = False
    tags: list[str] = Field(default_factory=list)


class DatasetAnnotationItem(BaseModel):
    path: str
    label_path: str | None = None
    format: str = "YOLO"
    annotations: list[DatasetAnnotationShape] = Field(default_factory=list)
    verified: bool = False
    tags: list[str] = Field(default_factory=list)


class DatasetTagCatalogPayload(BaseModel):
    tags: list[str] = Field(default_factory=list)


class DatasetTagCatalogItem(BaseModel):
    tags: list[str] = Field(default_factory=list)


class DatasetClassItem(BaseModel):
    id: str
    name: str
    instances: int
    color: str


class DatasetClassCreate(BaseModel):
    name: str = Field(min_length=1, max_length=120)


class DatasetClassesImport(BaseModel):
    classes: list[str] = Field(default_factory=list)
    mode: str = "append"


class DatasetSplitItem(BaseModel):
    name: str
    images: int
    videos: int
    total: int
    percent: float
    description: str
    color: str
    system: bool = False


class DatasetSplitCreate(BaseModel):
    name: str = Field(min_length=1, max_length=80)
    description: str = ""


class DatasetEventItem(BaseModel):
    id: int
    dataset_id: int
    event_type: str
    title: str
    description: str
    author: str
    created_at: datetime


class DatasetSourceItem(BaseModel):
    id: int
    dataset_id: int
    name: str
    source_type: str
    source_path: str
    target_path: str
    status: str
    split_policy: str
    images: int
    videos: int
    annotations: int
    classes: int
    frames: int
    train: int
    val: int
    test: int
    notes: str
    created_at: datetime
    updated_at: datetime


class DatasetSourceCreate(BaseModel):
    source_path: str = Field(min_length=1)
    name: str | None = None
    source_type: str = "folder"
    split_policy: str = "Auto"
    copy_assets: bool = True
    extract_video_frames: bool = False
    frame_interval: int = Field(default=30, ge=1)
    notes: str = ""


class DatasetSourceAddSummary(BaseModel):
    dataset: DatasetItem
    source: DatasetSourceItem
    sources: list[DatasetSourceItem]
    warnings: list[str] = Field(default_factory=list)


class DatasetVideoPlanRequest(BaseModel):
    source_path: str = Field(min_length=1)
    frame_interval: int = Field(default=30, ge=1)
    split_policy: str = "Auto"


class DatasetVideoPlanItem(BaseModel):
    name: str
    path: str
    size_bytes: int
    split: str
    duration_seconds: float | None = None
    fps: float | None = None
    total_frames: int | None = None
    estimated_frames: int
    warning: str = ""


class DatasetVideoPlanSummary(BaseModel):
    source_path: str
    video_count: int
    total_size_bytes: int
    total_duration_seconds: float | None = None
    estimated_frames: int
    frame_interval: int
    split_policy: str
    items: list[DatasetVideoPlanItem]
    warnings: list[str] = Field(default_factory=list)


class DatasetFrameExtractionRequest(BaseModel):
    frame_interval: int = Field(default=30, ge=1)


class DatasetFrameExtractionSummary(BaseModel):
    dataset: DatasetItem
    source: DatasetSourceItem
    sources: list[DatasetSourceItem]
    frames_saved: int
    warnings: list[str] = Field(default_factory=list)


class DatasetUploadSummary(BaseModel):
    dataset: DatasetItem
    files_saved: int
    images_saved: int
    videos_saved: int
    frames_saved: int
    assigned_splits: list[str]
    warnings: list[str] = Field(default_factory=list)


class DatasetMediaBulkPayload(BaseModel):
    paths: list[str] = Field(default_factory=list)
    operation: str = Field(min_length=1)
    split: str | None = None
    tags: list[str] = Field(default_factory=list)


class DatasetMediaBulkSummary(BaseModel):
    dataset: DatasetItem
    changed_files: int
    removed_from_version: int = 0
    moved_files: int = 0
    tagged_files: int = 0
    target_split: str = ""
    tags: list[str] = Field(default_factory=list)
    message: str = ""


class DatasetVersionCreate(BaseModel):
    version: str | None = None
    description: str = ""


class DatasetVersionSampleItem(BaseModel):
    path: str
    name: str
    kind: str
    split: str
    annotated: bool
    verified: bool


class DatasetVersionSummary(BaseModel):
    dataset: DatasetItem
    source_count: int
    has_manifest: bool
    manifest_path: str
    tracked_files: int
    missing_files: int
    image_files: int
    video_files: int
    annotated_files: int
    verified_files: int
    sample_files: list[DatasetVersionSampleItem] = Field(default_factory=list)


class DatasetExportSummary(BaseModel):
    dataset_id: int
    version: str
    export_format: str
    export_path: str
    images_exported: int
    labels_exported: int
    skipped_videos: int
    classes: int
    data_yaml_path: str
    classes_path: str


class DatasetCreate(BaseModel):
    name: str = Field(min_length=1, max_length=160)
    version: str = "v1.0"
    topic: str = "Custom"
    task: str = "Object Detection"
    format: str = "YOLO"
    source_path: str | None = None
    project_id: int | None = None


class DatasetImport(BaseModel):
    source_path: str = Field(min_length=1)
    name: str | None = None
    version: str = "v1.0"
    topic: str = "Custom"
    task: str = "Object Detection"
    format: str = "YOLO"
    project_id: int | None = None


class DatasetUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=160)
    version: str | None = None
    topic: str | None = None
    task: str | None = None
    status: str | None = None
    format: str | None = None
    source_path: str | None = None
    storage_path: str | None = None
    project_id: int | None = None


class ProjectSummary(BaseModel):
    id: int
    name: str
    description: str
    task: str
    status: str
    owner: str
    storage_path: str
    created_at: datetime
    updated_at: datetime
    dataset_count: int
    model_count: int
    experiment_count: int
    pipeline_count: int
    image_count: int
    video_count: int
    class_count: int
    storage_size_bytes: int


class ProjectActivityItem(BaseModel):
    id: int
    dataset_id: int
    dataset_name: str
    event_type: str
    title: str
    description: str
    created_at: datetime


class ProjectDetail(ProjectSummary):
    datasets: list[DatasetItem]
    activities: list[ProjectActivityItem] = Field(default_factory=list)


class ProjectCreate(BaseModel):
    name: str = Field(min_length=1, max_length=160)
    description: str = ""
    task: str = "Object detection and classification"
    storage_path: str = ""


class ProjectUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=160)
    description: str | None = None
    task: str | None = None
    status: str | None = None
    owner: str | None = None
    storage_path: str | None = None

File: frontend/src/types/workspace.ts
export type DatasetStats = {
  images: number;
  videos: number;
  annotations: number;
  classes: number;
  labeled: number;
  verified: number;
  labeled_percent: number;
  verified_percent: number;
};

export type DatasetSplits = {
  train: number;
  val: number;
  test: number;
};

export type DatasetItem = {
  id: number;
  project_id: number | null;
  name: string;
  version: string;
  topic: string;
  task: string;
  status: string;
  format: string;
  source_path: string;
  storage_path: string;
  created_at: string;
  updated_at: string;
  stats: DatasetStats;
  splits: DatasetSplits;
};

export type DatasetMediaItem = {
  id: string;
  name: string;
  path: string;
  kind: "image" | "video";
  split: "Train" | "Val" | "Test" | string;
  annotated: boolean;
  verified: boolean;
  tags: string[];
  preview_url: string | null;
  width: number | null;
  height: number | null;
};

export type DatasetAnnotationShape = {
  class_id: number;
  x_center: number;
  y_center: number;
  width: number;
  height: number;
};

export type DatasetAnnotationItem = {
  path: string;
  label_path: string | null;
  format: "YOLO" | string;
  annotations: DatasetAnnotationShape[];
  verified: boolean;
  tags: string[];
};

export type DatasetTagCatalogItem = {
  tags: string[];
};

export type DatasetClassItem = {
  id: string;
  name: string;
  instances: number;
  color: string;
};

export type DatasetSplitItem = {
  name: string;
  images: number;
  videos: number;
  total: number;
  percent: number;
  description: string;
  color: string;
  system: boolean;
};

export type DatasetEventItem = {
  id: number;
  dataset_id: number;
  event_type: string;
  title: string;
  description: string;
  author: string;
  created_at: string;
};

export type DatasetSourceItem = {
  id: number;
  dataset_id: number;
  name: string;
  source_type: string;
  source_path: string;
  target_path: string;
  status: string;
  split_policy: string;
  images: number;
  videos: number;
  annotations: number;
  classes: number;
  frames: number;
  train: number;
  val: number;
  test: number;
  notes: string;
  created_at: string;
  updated_at: string;
};

export type DatasetSourceCreatePayload = {
  source_path: string;
  name?: string;
  source_type?: string;
  split_policy?: string;
  copy_assets?: boolean;
  extract_video_frames?: boolean;
  frame_interval?: number;
  notes?: string;
};

export type DatasetSourceAddSummary = {
  dataset: DatasetItem;
  source: DatasetSourceItem;
  sources: DatasetSourceItem[];
  warnings: string[];
};

export type DatasetVideoPlanItem = {
  name: string;
  path: string;
  size_bytes: number;
  split: string;
  duration_seconds: number | null;
  fps: number | null;
  total_frames: number | null;
  estimated_frames: number;
  warning: string;
};

export type DatasetVideoPlanSummary = {
  source_path: string;
  video_count: number;
  total_size_bytes: number;
  total_duration_seconds: number | null;
  estimated_frames: number;
  frame_interval: number;
  split_policy: string;
  items: DatasetVideoPlanItem[];
  warnings: string[];
};

export type DatasetFrameExtractionSummary = {
  dataset: DatasetItem;
  source: DatasetSourceItem;
  sources: DatasetSourceItem[];
  frames_saved: number;
  warnings: string[];
};

export type DatasetUploadSummary = {
  dataset: DatasetItem;
  files_saved: number;
  images_saved: number;
  videos_saved: number;
  frames_saved: number;
  assigned_splits: string[];
  warnings: string[];
};

export type DatasetMediaBulkPayload = {
  paths: string[];
  operation: string;
  split?: string | null;
  tags?: string[];
};

export type DatasetMediaBulkSummary = {
  dataset: DatasetItem;
  changed_files: number;
  removed_from_version: number;
  moved_files: number;
  tagged_files: number;
  target_split: string;
  tags: string[];
  message: string;
};

export type DatasetImportPayload = {
  source_path: string;
  name?: string;
  version?: string;
  topic?: string;
  task?: string;
  format?: string;
  project_id?: number | null;
};

export type DatasetCreatePayload = {
  name: string;
  version?: string;
  topic?: string;
  task?: string;
  format?: string;
  source_path?: string;
  project_id?: number | null;
};

export type DatasetUpdatePayload = Partial<{
  name: string;
  version: string;
  topic: string;
  task: string;
  status: string;
  format: string;
  source_path: string;
  storage_path: string;
  project_id: number | null;
}>;

export type DatasetVersionCreatePayload = {
  version?: string;
  description?: string;
};

export type DatasetVersionSampleItem = {
  path: string;
  name: string;
  kind: "image" | "video" | string;
  split: string;
  annotated: boolean;
  verified: boolean;
};

export type DatasetVersionSummary = {
  dataset: DatasetItem;
  source_count: number;
  has_manifest: boolean;
  manifest_path: string;
  tracked_files: number;
  missing_files: number;
  image_files: number;
  video_files: number;
  annotated_files: number;
  verified_files: number;
  sample_files: DatasetVersionSampleItem[];
};

export type DatasetExportSummary = {
  dataset_id: number;
  version: string;
  export_format: string;
  export_path: string;
  images_exported: number;
  labels_exported: number;
  skipped_videos: number;
  classes: number;
  data_yaml_path: string;
  classes_path: string;
};

export type ProjectSummary = {
  id: number;
  name: string;
  description: string;
  task: string;
  status: string;
  owner: string;
  storage_path: string;
  created_at: string;
  updated_at: string;
  dataset_count: number;
  model_count: number;
  experiment_count: number;
  pipeline_count: number;
  image_count: number;
  video_count: number;
  class_count: number;
  storage_size_bytes: number;
};

export type ProjectDetail = ProjectSummary & {
  datasets: DatasetItem[];
  activities: ProjectActivityItem[];
};

export type ProjectCreatePayload = {
  name: string;
  description?: string;
  task?: string;
  storage_path?: string;
};

export type ProjectUpdatePayload = Partial<{
  name: string;
  description: string;
  task: string;
  status: string;
  owner: string;
  storage_path: string;
}>;

export type ProjectActivityItem = {
  id: number;
  dataset_id: number;
  dataset_name: string;
  event_type: string;
  title: string;
  description: string;
  created_at: string;
};
